mips/exc.S

```diff
@@ -70,9 +71,19 @@ user_exc_enter:
         SAVE_REG($t2, BADVADDR, $k0)
         SAVE_REG($t3, PC, $k0)
 
-        # TODO: switch to kernel SP!
+        # Load kernel global pointer.
         la      $gp, _gp
 
+        # Load kernel stack pointer (always stack bottom ?)
```

Comment:
I've been thinking about this, and I believe that it is a safe assumption. Here's how it works:

We restore the kernel `$sp` only when entering interrupt handler while a user-mode code executes, which is only the case for user threads that are executing user code. Kernel threads don't need this, because their correct stack pointer is already stored in their saved context. User threads that are in progress of handling another interrupt also don't need this, for the same reason. But user threads normally work on their dedicated stack in user memory, so we need to temporarily move them to a safe stack somewhere in kernel memory. There are exactly two scenarios when this happens:

- Either this interrupt we are handling is **not** the first interrupt this user thread experienced. In that case, what could be left on the kernel stack? The last thing it was used for was processing that previous interrupt, but, apparently, that last time we have returned to user mode (via `user_exc_leave`), so the entire stack that was used for handling that interrupt has rolled back, and is now empty. Since no data is left on the stack, we can safely restore `$sp` to kernel stack bottom.

- The other case is when this is the very first interrupt this thread receives after switching to user-mode. In that case kernel-stack was most recently used to perform the initial switch to user mode, with `fork_trampoline` or similar procedure. However, `fork_trampoline` is not expected to return, so it's caller (and the caller's caller etc.) do not expect to ever continue execution. Therefore, in this case, even though the stack is non-empty and does contain stored variables, they are no longer relevant and can never be accessed anymore, so it is safe to throw them away and treat the kernel stack as empty. So again, we can start at stack bottom.

By induction, it is always safe to restore `$sp` to kernel stack bottom - when the interrupt is received in user mode.

Note that these observations are unaffected by diverging contexts: if a new kernel thread is created, it starts with an empty kernel stack; if a user thread `fork()s`, it begins user-mode execution with kernel stack being used for `fork_trampoline`.

Also, the proposed behavior is compatible with recursive interrupts. The first interrupt received may require switching from user stack to kernel stack, but if a subsequent interrupt is received before we are done handling this one, we are already in kernel-mode, so the `$sp` won't be moved to kernel stack (as it already points somewhere inside it), and will be correctly used to recursively handle the new interrupt.

Maybe it would be a good idea to explain that briefly in a comment in sources.